packages/backend-core/src/Endpoint/Endpoint.ts

```diff
@@ -0,0 +1,74 @@
+import Router from "@koa/router"
+import { UserCtx, Ctx } from "@budibase/types"
+import { Next } from "koa"
+
+export type Method = "post" | "put" | "patch" | "get" | "delete" | "head"
+
+export type UserCtxWithNextFn = (
+  ctx: UserCtx,
+  next: Next
+) => void | Promise<unknown>
+export type CtxWithNextFn = (ctx: Ctx, next: Next) => void | Promise<unknown>
+export type UserCtxWithoutNextFn = (ctx: UserCtx) => void | Promise<unknown>
+export type CtxWithoutNextFn = (ctx: Ctx) => void | Promise<unknown>
+export type CtxFn =
+  | UserCtxWithNextFn
+  | UserCtxWithoutNextFn
+  | CtxWithNextFn
+  | CtxWithoutNextFn
+
+class Endpoint {
+  method: Method
+  url: string
+  controller: CtxFn
+  middlewares: CtxFn[]
+  outputMiddlewares: CtxFn[]
+
+  constructor(method: Method, url: string, controller: CtxFn) {
+    this.method = method
+    this.url = url
+    this.controller = controller
+    this.middlewares = []
+    this.outputMiddlewares = []
+  }
+
+  addMiddleware(middleware: CtxFn, opts?: { first?: boolean }) {
+    if (opts?.first) {
```

Comment:
Why using this `first` flag instead of actually relying on the addition order?

Reply:
the group middlewares are independent from the middlewares that have been added to the endpoint directly, generally the pre-middlewares will always come first and the post-middlewares will always come last. I wanted to provide a way to flip this behaviour if needs be, this is needed for the authorization middlewares, they come after the resource IDs have been picked out. I wanted a way to still apply these to the group, but to specify they should be added after the endpoint middlewares.